include/rcutils/shared_library.h

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCUTILS__SHARED_LIBRARY_H_
+#define RCUTILS__SHARED_LIBRARY_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <string.h>
+
+#ifndef _WIN32
+#include <dlfcn.h>
+typedef void * rcutils_shared_library_handle_t;
+#else
+#include <windows.h>
+typedef HINSTANCE rcutils_shared_library_handle_t;
+#endif  // _WIN32
+
+#include "rcutils/allocator.h"
+#include "rcutils/types/rcutils_ret.h"
+#include "rcutils/macros.h"
+#include "rcutils/visibility_control.h"
+
+/// Handle to a loaded shared library.
+typedef struct RCUTILS_PUBLIC_TYPE rcutils_shared_library_t
+{
+  /// The pointer to the shared library
+  rcutils_shared_library_handle_t lib_pointer;
+  /// The path of the shared_library
+  char * library_path;
+  /// allocator
+  rcutils_allocator_t allocator;
+} rcutils_shared_library_t;
+
+/// Return an empty shared library struct.
+/*
+ * This function returns an empty and zero initialized shared library struct.
+ *
+ * Example:
+ *
+ * ```c
+ * // Do not do this:
+ * // rcutils_shared_library_t foo;
+ * // rcutils_load_shared_library(
+ * //     &foo,
+ * //    "library_name",
+ * //    rcutils_get_default_allocator()); // undefined behavior!
+ * // or
+ * // rcutils_unload_shared_library(&foo); // undefined behavior!
+ *
+ * // Do this instead:
+ * rcutils_shared_library_t bar = rcutils_get_zero_initialized_shared_library();
+ * rcutils_load_shared_library(&bar, "library_name",rcutils_get_default_allocator()); // ok
+ * void * symbol = rcutils_get_symbol(&bar, "bazinga"); // ok
+ * bool is_bazinga_symbol = rcutils_has_symbol(&bar, "bazinga"); // ok
+ * rcutils_unload_shared_library(&bar); // ok
```

Comment:
You should show the capture and checking of the return type from `rcutils_unload_shared_library()`.